vllm/model_executor/model_loader/loader.py

```
@@ -526,6 +533,201 @@ def save_model(
+            hf_weights_files = glob.glob(
+                os.path.join(hf_folder, matched_pattern))
+        else:
+            hf_folder = model_name_or_path
+            for pattern in allowed_patterns:
+                hf_weights_files = glob.glob(os.path.join(hf_folder, pattern))
+                if len(hf_weights_files) > 0:
+                    matched_pattern = pattern
+                    break
+
+        if matched_pattern != "*.safetensors":
+            hf_weights_files = filter_files_not_needed_for_inference(
+                hf_weights_files)
+
+        if len(hf_weights_files) == 0:
+            raise RuntimeError(
+                f"Cannot find any model weights with `{model_name_or_path}`")
+
+        return hf_weights_files, matched_pattern == "*.safetensors"
+
+    def _get_quantized_weights_iterator(
+        self, model_name_or_path: str, revision: Optional[str]
+    ) -> Tuple[Generator[Tuple[str, torch.Tensor], None, None], Dict[str,
+                                                                     Any]]:
+        """Get an iterator for the model weights before bitsandbytes quantization and return the quantization state dictionary."""
+        hf_weights_files, use_safetensors = self._prepare_weights(
+            model_name_or_path, revision)
+
+        quant_state_dict = {}
+        if use_safetensors:
+            weight_iterator = safetensors_weights_iterator(hf_weights_files)
+        else:
+            weight_iterator = pt_weights_iterator(hf_weights_files)
+        current_tensor_type = torch.get_default_dtype()
+
+        def generator():
+            for weight_name, weight_tensor in weight_iterator:
+                if any(target_module in weight_name
+                       for target_module in self.target_modules):
+                    weight_name = weight_name.replace(".weight", ".qweight")
+                    loaded_weight = weight_tensor.cuda().data  # Move to GPU
```

move to device instead? if bnb only supports cuda, we should throw an exception much earlier if the device doesn't match (remember vllm supports multiple accelerator types)

---

bnb quantization requires the data to be quantized to be in GPU. The weight_tensor are the weights just read from the weight files downloaded from hugging face, so it can only be in CPU. Therefore, I explicitly have to move them GPU explicitly.